soco/events_base.py

```diff
@@ -0,0 +1,642 @@
+    `soco.events_twisted.EventListener`.
+    """
+
+    def __init__(self):
+        #: `bool`: Indicates whether the server is currently running
+        self.is_running = False
+        self._start_lock = threading.Lock()
+        #: `tuple`: The address (ip, port) on which the server is
+        #: configured to listen.
+        # Empty for the moment. (It is set in `start`)
+        self.address = ()
+        #: `int`: Port on which to listen.
+        self.requested_port_number = config.EVENT_LISTENER_PORT
+
+    def start(self, any_zone):
+        """Start the event listener listening on the local machine.
+
+        Args:
+            any_zone (SoCo): Any Sonos device on the network. It does not
+                matter which device. It is used only to find a local IP
+                address reachable by the Sonos net.
+
+        Note:
+            This method calls the listen method, which must be overridden
+            in the class that inherits from this class.
+        """
+
+        # Find our local network IP address which is accessible to the
+        # Sonos net, see http://stackoverflow.com/q/166506
+        with self._start_lock:
+            if not self.is_running:
+                # Use configured IP address if there is one, else detect
+                # automatically.
+                if config.EVENT_LISTENER_IP:
+                    ip_address = config.EVENT_LISTENER_IP
+                else:
+                    temp_sock = socket.socket(socket.AF_INET,
+                                              socket.SOCK_DGRAM)
+                    try:
+                        # doesn't have to be reachable
+                        temp_sock.connect((any_zone.ip_address, 0))
+                        ip_address = temp_sock.getsockname()[0]
+                    except socket.error:
+                        log.exception(
+                            'Could not start Event Listener: check network.')
+                        ip_address = None
+                    finally:
+                        temp_sock.close()
+                if ip_address:  # Otherwise, no point trying to start server
+                    # Check what port we actually got (twisted only)
+                    port = self.listen(ip_address)
+                    if port:
+                        self.address = (ip_address, port)
+                        self.is_running = True
+                        log.info("Event Listener started")
+
+    def stop(self):
+        """Stop the Event Listener.
+
+        Note:
+            This method calls the stop_listening method, which must be
+            overridden in the class that inherits from this class.
+        """
+        if not self.is_running:
+            return
+        self.is_running = False
+        self.stop_listening(self.address)
+        log.info("Event Listener stopped")
+
+    # pylint: disable=missing-docstring
+    def listen(self, ip_address):
+        pass
```

Comment:
By convention, abstract methods should raise NotImplementedException.

Reply:
Will do.